src/OVAL/probes/unix/linux/selinuxboolean_probe.c

```diff
@@ -44,28 +44,154 @@
 
 #include "_seap.h"
 #include "probe-api.h"
 #include "probe/entcmp.h"
+#include <probe/probe.h>
 
 #include <selinux/selinux.h>
 #include "selinuxboolean_probe.h"
 
+
+/* These definitions and three functions were commandeered from SELinux internals
+ * and creatively improved to take the global prefix into account */
+
+#define OLDSELINUXMNT "/selinux"
+#define SELINUXMNT    "/sys/fs/selinux"
+#define SELINUXFS     "selinuxfs"
+#define SELINUX_MAGIC 0xf97cff8c
+
+static int verify_selinuxmnt_prefixed(const char *prefix, const char *mnt)
+{
+	char path[PATH_MAX] = {0};
+	struct statfs sfbuf;
+	int rc;
+
+	if (!prefix)
+		return -1;
+
+	snprintf(path, PATH_MAX, "%s%s", prefix, mnt);
+
+	do {
+		rc = statfs(path, &sfbuf);
+	} while (rc < 0 && errno == EINTR);
+	if (rc == 0) {
+		if ((uint32_t)sfbuf.f_type == (uint32_t)SELINUX_MAGIC) {
+			struct statvfs vfsbuf;
+			rc = statvfs(path, &vfsbuf);
+			if (rc == 0) {
+				if (!(vfsbuf.f_flag & ST_RDONLY)) {
+					set_selinuxmnt(path);
+				}
+				return 0;
+			}
+		}
+	}
+
+	return -1;
+}
+
+static int selinuxfs_exists_prefixed(const char *prefix)
+{
+	char path[PATH_MAX] = {0};
+	int exists = 0;
+	FILE *fp = NULL;
+	char *buf = NULL;
+	size_t len;
+	ssize_t num;
+
+	if (!prefix)
+		return 0;
+
+	snprintf(path, PATH_MAX, "%s/proc/filesystems", prefix);
+	fp = fopen(path, "re");
+	if (!fp)
+		return 1; /* Fail as if it exists */
+	__fsetlocking(fp, FSETLOCKING_BYCALLER);
+
+	num = getline(&buf, &len, fp);
+	while (num != -1) {
+		if (strstr(buf, SELINUXFS)) {
+			exists = 1;
+			break;
+		}
+		num = getline(&buf, &len, fp);
+	}
+
+	free(buf);
+	fclose(fp);
+	return exists;
+}
+
+static void init_selinuxmnt_prefixed(const char *prefix)
+{
+	char path[PATH_MAX] = {0};
+	char *buf = NULL, *p;
+	FILE *fp = NULL;
+	size_t len;
+	ssize_t num;
+
+	if (!prefix)
+		return;
+
+	if (verify_selinuxmnt_prefixed(prefix, SELINUXMNT) == 0) return;
+
+	if (verify_selinuxmnt_prefixed(prefix, OLDSELINUXMNT) == 0) return;
+
+	/* Drop back to detecting it the long way. */
+	if (!selinuxfs_exists_prefixed(prefix))
+		goto out;
+
+	/* At this point, the usual spot doesn't have an selinuxfs so
+	 * we look around for it */
+	snprintf(path, PATH_MAX, "%s/proc/mounts", prefix);
+	fp = fopen(path, "re");
+	if (!fp)
+		goto out;
+
+	__fsetlocking(fp, FSETLOCKING_BYCALLER);
+	while ((num = getline(&buf, &len, fp)) != -1) {
+		char *tmp;
+		p = strchr(buf, ' ');
+		if (!p)
+			goto out;
+		p++;
+		tmp = strchr(p, ' ');
+		if (!tmp)
+			goto out;
+		if (!strncmp(tmp + 1, SELINUXFS" ", strlen(SELINUXFS)+1)) {
+			*tmp = '\0';
+			break;
+		}
+	}
+
+	/* If we found something, dup it */
+	if (num > 0)
+		verify_selinuxmnt_prefixed(prefix, p);
+
+out:
+	free(buf);
+	if (fp)
+		fclose(fp);
+	return;
+}
+
+/* End of borrowing from SELinux */
+
 static int get_selinuxboolean(SEXP_t *ut_ent, probe_ctx *ctx)
 {
 	int err = 1, active, pending, len, i;
 	SEXP_t *boolean, *item;
 	char **booleans;
 
-	if ( ! is_selinux_enabled()) {
```

Comment:
What will happen now if SELinux isn't enabled?

Reply:
SElinux might be disabled in the container where the scanner is running, but enabled on the host. In this situation we still can check SELinux booleans in `/host/sys/fs/selinux`.

Moreover I don't really understand why this check was needed in the first place. If there is no mounted selinux filesystem, the probe will return an error. If there is — it will fetch values.